python/ray/dag/dag_node_operation.py

```diff
@@ -384,140 +342,103 @@ def _select_next_nodes(
-            the inner list, the order of operations is READ, COMPUTE, and WRITE.
+        actor_to_operation_nodes: A dictionary that maps an actor handle to a list of
+            _DAGOperationGraphNodes. For the same actor, the index in the list
+            corresponds to the index of the ExecutableTask in the list of
+            `executable_tasks` in `actor_to_executable_tasks`.
 
     Returns:
-        A graph where each node is a _DAGOperationGraphNode. The key is `task_idx`,
-        the index to retrieve its task from `idx_to_task`, and the value is a
-        dictionary that maps the _DAGNodeOperationType (READ, COMPUTE, or WRITE)
-        to the corresponding _DAGOperationGraphNode
+        A dictionary mapping a task index to its corresponding _DAGOperationGraphNode.
     """
-    assert idx_to_task
-    graph: Dict[int, Dict[_DAGNodeOperationType, _DAGOperationGraphNode]] = {}
-
-    for _, operation_nodes_list in actor_to_operation_nodes.items():
-        prev_compute_node = None
-        for operation_nodes in operation_nodes_list:
-            task_idx = operation_nodes[0].task_idx
-            read_node, compute_node, write_node = (
-                operation_nodes[0],
-                operation_nodes[1],
-                operation_nodes[2],
-            )
-            # Add edges from READ to COMPUTE, and from COMPUTE to WRITE, which
-            # belong to the same task.
-            _add_edge(read_node, compute_node)
-            _add_edge(compute_node, write_node)
-            # Add an edge from COMPUTE with `bind_index` i to COMPUTE with
-            # `bind_index` i+1 if they belong to the same actor.
-            if prev_compute_node is not None:
-                _add_edge(prev_compute_node, compute_node, "", True)
-            prev_compute_node = compute_node
-            assert task_idx not in graph
-            graph[task_idx] = {
-                _DAGNodeOperationType.READ: read_node,
-                _DAGNodeOperationType.COMPUTE: compute_node,
-                _DAGNodeOperationType.WRITE: write_node,
-            }
-
-    # Import `ray.dag` here to avoid circular import.
     from ray.dag import ClassMethodNode, CollectiveOutputNode, MultiOutputNode
-    from ray.dag.collective_node import _CollectiveOperation
-
-    # Add an edge from WRITE of the writer task to READ of the reader task.
+    from ray.dag.communication_node import _CollectiveOperation
+
+    graph: Dict[int, _DAGOperationGraphNode] = {}
+    # Add control edges between tasks from the same actor.
+    for op_nodes in actor_to_operation_nodes.values():
+        for i, node in enumerate(op_nodes):
+            assert node.task_idx not in graph
+            graph[node.task_idx] = node
+            if i > 0:
+                prev_node = op_nodes[i - 1]
+                # Skip adding a control dependency edge for the created NCCL P2P
+                # send/recv nodes, which have the same task index as the previous node.
```

Comment:
Does this work because send/recv nodes have a data dependency on the previous node? If so, please make a comment about that.

Reply:
Yes. Done.